OutOfSchool/OutOfSchool.Redis/CacheService.cs

```diff
@@ -0,0 +1,130 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using StackExchange.Redis;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis
+{
+    public class CacheService : ICacheService, IDisposable
+    {
+        private readonly IDistributedCache cache;
+        private readonly RedisConfig redisConfig;
+
+        private bool isWorking = true;
+        private readonly bool isEnabled = false;
+        private Timer timer;
+
+        public CacheService(
+            IDistributedCache cache, 
+            IOptions<RedisConfig> redisConfig
+        )
+        {
+            this.cache = cache;
+            
+            try
+            {
+                this.redisConfig = redisConfig.Value;
+                isEnabled = this.redisConfig.Enabled;
+            }
+            catch (OptionsValidationException)
+            {
+                isEnabled = false;
+            }
+        }
+
+        public async Task<T> Get<T>(string key)
+        {
+            try
+            {
+                if (isEnabled && isWorking)
+                {
+                    var value = await cache.GetStringAsync(key);
+
+                    if (value != null)
+                    {
+                        return JsonConvert.DeserializeObject<T>(value);
+                    }
+                }
+            }
+            catch (RedisConnectionException)
+            {
+                RedisIsBrokenStartCheck();
+            }
+
+            return default;
+        }
+
+        public async Task Set<T>(string key, T value)
+        {
+            await ExecuteRedisMethod(async () => {
+                var options = new DistributedCacheEntryOptions
+                {
+                    AbsoluteExpirationRelativeToNow = TimeSpan.FromHours(redisConfig.AbsoluteExpirationRelativeToNowFromHours),
+                    SlidingExpiration = TimeSpan.FromMinutes(redisConfig.SlidingExpirationFromMinutes)
+                };
+
+                await cache.SetStringAsync(key, JsonConvert.SerializeObject(value), options);
+            });
+        }
+
+        public async Task ClearCache(string key)
+        {
+            await ExecuteRedisMethod(async () => {
+                await cache.RemoveAsync(key);
+            });
+        }
+
+        public async Task Refresh(string key)
+        {
+            await ExecuteRedisMethod(async () => {
+                await cache.RefreshAsync(key);
+            });
+        }
+
+        public void Dispose()
+        {
+            timer?.Dispose();
+        }
+
+        private void RedisIsBrokenStartCheck()
+        {
+            isWorking = false;
+
+            timer = new Timer(
+                cb =>
+                {
+                    try
+                    {
+                        cache.GetString("_");
```

Comment:
Variable lockFlagRedisIsBroken was added to prevent unnecessary calls. Class Interlocked was used.

Reply:
I'm afraid the issue is still not fixed. The `lockFlagRedisIsBroken` (extra credits for using `int` and naming it `...flag` :) ) is not checked when timer callback occurs. 
1. Thus we could have many timer events occurred in parallel if `cache.GetString("_");` is long enough
2. They could decrease `lockFlagRedisIsBroken` more than once, and you'll get < 0 value there, and your condition in 
`if (Interlocked.CompareExchange(ref lockFlagRedisIsBroken, 1, 0) == 0)`
would never occur, making `RedisIsBrokenStartCheck()` be noop till app restart.

There are several solutions to this problem, you could use something simple like `Monitor.TryEnter()` for the whole method and simple cycle and `await Task.Delay()` instead of timer (this way you could also easily specify max retry count or max duration of polling and cancellation token if you need it, whatever)

Reply:
Used Monitor and Task.Delay